File: ios/Auth/UserSession.swift
import Foundation

/// Persists JWT token and user profile in UserDefaults.
/// Keeps ApiClient.shared.authToken in sync.
final class UserSession: ObservableObject {
    static let shared = UserSession()
    private init() { load() }

    private let defaults = UserDefaults.standard

    @Published var isLoggedIn = false
    @Published var profile: UserProfile?

    // MARK: Public API

    func save(token: String, profile: UserProfile) {
        defaults.set(token,                     forKey: "token")
        defaults.set(profile.id,               forKey: "user_id")
        defaults.set(profile.email,            forKey: "email")
        defaults.set(profile.vehicleType,      forKey: "vehicle_type")
        defaults.set(profile.isDisabled,       forKey: "is_disabled")
        defaults.set(profile.hasResidentPermit, forKey: "has_resident_permit")
        defaults.set(profile.residentZone,     forKey: "resident_zone")

        ApiClient.shared.authToken = token
        self.profile = profile
        self.isLoggedIn = true
    }

    func updateProfile(_ updated: UserProfile) {
        guard let token = defaults.string(forKey: "token") else { return }
        save(token: token, profile: updated)
    }

    func logout() {
        ["token", "user_id", "email", "vehicle_type",
         "is_disabled", "has_resident_permit", "resident_zone"].forEach {
            defaults.removeObject(forKey: $0)
        }
        ApiClient.shared.authToken = nil
        profile = nil
        isLoggedIn = false
    }

    // MARK: Private

    private func load() {
        guard let token = defaults.string(forKey: "token") else { return }
        ApiClient.shared.authToken = token
        profile = UserProfile(
            id: defaults.integer(forKey: "user_id"),
            email: defaults.string(forKey: "email") ?? "",
            vehicleType: defaults.string(forKey: "vehicle_type") ?? "car",
            isDisabled: defaults.bool(forKey: "is_disabled"),
            hasResidentPermit: defaults.bool(forKey: "has_resident_permit"),
            residentZone: defaults.string(forKey: "resident_zone") ?? ""
        )
        isLoggedIn = true
    }
}

File: android/app/src/main/java/com/parksence/auth/UserSession.kt
package com.parksence.auth

import android.content.Context
import com.parksence.api.ApiClient
import com.parksence.api.UserProfile

/**
 * Persists JWT token and user profile in SharedPreferences.
 * Also keeps ApiClient.authToken in sync.
 */
object UserSession {

    private const val PREFS = "session"

    fun save(context: Context, token: String, profile: UserProfile) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().apply {
            putString("token", token)
            putInt("user_id", profile.id)
            putString("email", profile.email)
            putString("vehicle_type", profile.vehicleType)
            putBoolean("is_disabled", profile.isDisabled)
            putBoolean("has_resident_permit", profile.hasResidentPermit)
            putString("resident_zone", profile.residentZone)
            apply()
        }
        ApiClient.authToken = token
    }

    fun load(context: Context): UserProfile? {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val token = prefs.getString("token", null) ?: return null
        ApiClient.authToken = token
        return UserProfile(
            id = prefs.getInt("user_id", 0),
            email = prefs.getString("email", "") ?: "",
            vehicleType = prefs.getString("vehicle_type", "car") ?: "car",
            isDisabled = prefs.getBoolean("is_disabled", false),
            hasResidentPermit = prefs.getBoolean("has_resident_permit", false),
            residentZone = prefs.getString("resident_zone", "") ?: "",
        )
    }

    fun updateProfile(context: Context, profile: UserProfile) {
        val token = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString("token", null) ?: return
        save(context, token, profile)
    }

    fun isLoggedIn(context: Context): Boolean =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).contains("token")

    fun logout(context: Context) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().clear().apply()
        ApiClient.authToken = null
    }
}
